finality-aleph/src/sync/substrate/verification/cache.rs

```diff
@@ -0,0 +1,340 @@
+use std::{
+    collections::HashMap,
+    fmt::{Display, Error as FmtError, Formatter},
+};
+
+use aleph_primitives::BlockNumber;
+
+use crate::{
+    session::{first_block_of_session, session_id_from_block_num, SessionId},
+    session_map::AuthorityProvider,
+    sync::substrate::verification::{verifier::SessionVerifier, FinalizationStatus},
+    SessionPeriod,
+};
+
+/// Ways in which a justification can fail verification.
+#[derive(Debug, PartialEq, Eq)]
+pub enum CacheError {
+    UnknownSession,
+    SessionTooOld(SessionId, SessionId),
+    SessionInFuture(SessionId, SessionId),
+}
+
+impl Display for CacheError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use CacheError::*;
+        match self {
+            SessionTooOld(session, lower_bound) => write!(
+                f,
+                "justification from too old session {:?}. Should be at least {:?}",
+                session, lower_bound
+            ),
+            SessionInFuture(session, upper_bound) => write!(
+                f,
+                "justification from session {:?} without known authorities. Should be at most {:?}",
+                session, upper_bound
+            ),
+            UnknownSession => write!(f, "justification from unknown session"),
+        }
+    }
+}
+
+/// Cache storing SessionVerifier structs for multiple sessions. Keeps up to `cache_size` verifiers of top sessions.
+/// If the session is too new or ancient it will fail to return a SessionVerifier.
+/// Highest session verifier this cache returns is for the session after the current finalization session.
+/// Lowest session verifier this cache returns is for `top_returned_session` - `cache_size`.
+pub struct VerifierCache<AP, FS>
+where
+    AP: AuthorityProvider<BlockNumber>,
+    FS: FinalizationStatus,
+{
+    sessions: HashMap<SessionId, SessionVerifier>,
+    session_period: SessionPeriod,
+    finalization_status: FS,
+    authority_provider: AP,
+    cache_size: SessionId,
```

Comment:
teribble nit: this should be just usize or (`type SessionCount = usize`). We use this to count sessions kept in the cache

Reply:
Yeah, maybe it should. We should probably somehow systematize all such types, but right now almost nothing in `finality-aleph`, or `bin/node` is dependent on `aleph_primitives` types such as `BlockNumber`, `SessionIndex`, `SessionCount`. They are currently set separately in two different places. There is a plan to unify it in [A0-1839](https://cardinal-cryptography.atlassian.net/browse/A0-1839). For now I would leave it as it is. If `SessionId` can hold a session number, it can also hold a `SessionCount`.

Reply:
Still, I would prefer to not mix an id with a count. Using usize here would be fine

Reply:
Using usize would be inconvenient for all arithmetics here. Session id is already treated and used as session index, so seems better for arithmetics imo.

Reply:
Its just misleading that the sessionid is used to count something. Using usize should not be that big of a deal imho

Reply:
It is misleading and to do it right we should rename it to `SessionIndex`. We already use it to count something in several places. We use partial ord from `u32`. `SessionId`'s are generated by a sequence. Substrate assumes they are a sequence and are `u32`.

So I would go with renaming it. And also changing it to `aleph_primitives::SessionIndex` / deriving traits such ass `Add`, `Sub` and others. All this should be done in a future refactor.

Alternatively creating a new type `SessionIndex = aleph_primitives::SessionIndex`, using it here and changing `pub struct SessionId(SessionIndex);`. If we want it to still be and id, we can leave it as an id in all other places, but use `SessionIndex` wherever there is functionality depending on it being and Index (so here, sesssion_map.rs, session.rs, party.rs). (I could do it in this PR)

The thing I don't want to do is use `usize` or `u32` here, like in `session_map.rs`.

Reply:
w8, in most places you use `cache_size.0` anyway and other place is in fn `prune` which is called like that also
```rust
self.prune(SessionId(
                session_id.0.saturating_sub(self.cache_size.0) + 1,
            ));
```
, so I dont understand the problem here. Secondly, this variable indicates the size of the cache and has nothing to do with a session - so no shenanigans with SessionIndex and derivatives of it would make sense. Using usize should be natural here. My initial comment with using the type alias `SessionCount = usize` was to just indicate that we are counting sessions here.

Reply:
Ah, yeah, you are right. I can use:
```
self.lower_bound.0.saturating_add(self.cache_size.saturated_into())
```
I will add it to this PR.

Still, changing `SessionId` to `SessionIndex = aleph_primitives::SessionIndex/u32` would remove need of using `session_id.0`. This would simplify a lot of stuff. Pattern of wrapping basic types does not really make sense if you can just get the inside of this type.